[Clang] Fix name lookup for dependent bases

vbe-sc opened the pull request:
Currently the following example is a compilation failure: 
```cpp
template<typename T> struct A {
    typedef int M;
    struct B {
      typedef void M;
      struct C;
    };
};

template<typename T> struct A<T>::B::C : A<T> {
    M m; // void or int ?
};
```

According to the point 13.8.3.2

```
A dependent base class is a base class that is a dependent type and is not the current instantiation.
Note 2 : A base class can be the current instantiation in the case of a nested class naming an enclosing class as a base.
```

The base class `A` is the current instantiation, because `C` is a nested class for an enclosing class `A<T>`, it's is the not-dependent base class and we need to search the names through its scope.

This patch makes this example compile

vbe-sc committed:
[clang] Fix name lookup for dependent bases

vbe-sc commented:
@Endilll, please, could you take a look? 
Also, do you have any ideas why the Win pipeline fails with checkout failure? How can I fix it?

Endilll commented:
> Also, do you have any ideas why the Win pipeline fails with checkout failure? How can I fix it?

It seems to be unrelated to you. Might be fixed when you'll push a new commit.

vbe-sc commented:
> > @Endilll, please, could you take a look?
> 
> My knowledge of wording about templates is not deep enough, unfortunately.

Could you, please, suggest someone who can review this part?

Endilll commented:
> > > @Endilll, please, could you take a look?
> > 
> > 
> > My knowledge of wording about templates is not deep enough, unfortunately.
> 
> Could you, please, suggest someone who can review this part?

I've already added several people who should be able to, but mind that we're very short on reviewers, so this will likely take time.

mizvekov commented on clang/lib/AST/CXXInheritance.cpp:
```diff
@@ -169,14 +169,18 @@ bool CXXBasePaths::lookupInBases(ASTContext &Context,
     // Find the record of the base class subobjects for this type.
     QualType BaseType =
         Context.getCanonicalType(BaseSpec.getType()).getUnqualifiedType();
+    bool isCurrentInstantiation = false;
+    if (auto *TST = BaseSpec.getType()->getAs<TemplateSpecializationType>())
+      isCurrentInstantiation = TST->isCurrentInstantiation();
```
This relies on a TemplateSpecializationType which is a type sugar to make this determination, and type sugar could be lost, for example with template instantiation.

Might be worth double checking and adding a test case where the base type is formed from instantiation.

This test would be equivalent to checking if the type is canonically an InjectedClassNameType:
```suggestion
    bool isCurrentInstantiation = isa<InjectedClassNameType>(BaseType);
```
This is reliable as it only depends on the canonical type.

vbe-sc replied:
You are right, thanks.
I've accepted your changes

mizvekov commented on clang/lib/AST/CXXInheritance.cpp:
```diff
@@ -265,8 +268,8 @@ bool CXXBasePaths::lookupInBases(ASTContext &Context,
       } else {
-        BaseRecord = cast<CXXRecordDecl>(
-            BaseSpec.getType()->castAs<RecordType>()->getDecl());
+        if (auto *RT = BaseSpec.getType()->getAs<RecordType>())
+          BaseRecord = cast<CXXRecordDecl>(RT->getDecl());
```
This should be equivalent to:
```suggestion
        BaseRecord = BaseSpec.getType()->getAsRecordDecl();
```

vbe-sc replied:
This is almost same. To be correct, we need to replace the first variant with 
```c++
BaseRecord = cast<CXXRecordDecl>(BaseSpec.getType()->getAsRecordDecl());
```

All in all, thanks for this comment. Fixed

vbe-sc commented:
@sdkrystian, please, could you take a look?

mizvekov reviewed (approved):
LGTM, but give some time for @sdkrystian to take a look, since he is actively working on this area.

vbe-sc commented:
@sdkrystian, could you please take a look?

sdkrystian commented:
@vbe-sc I can later today or tomorrow.

vbe-sc commented:
@sdkrystian, could you, please, take a look? We are about to merge this

vbe-sc commented:
> Please wait for a week before pinging someone again.
> 
> I think this is fine to merge as is, if you want to go ahead. We can always do post commit review later.

Sorry, I didn't know.
Let's do it this way you want

sdkrystian reviewed (commented):
I don't think this patch fixes the following case:
```cpp
template<typename T> 
struct A 
{
    struct B 
    {
        using X = int;

        struct C
        {
            using X = void;

            struct D;
        };
    };
};

template<typename T> 
struct A<T>::B::C::D : B
{
    X x; // error: field has incomplete type 'X' (aka 'void')
};
```
Regardless, this patch should probably include this as a test.

kolobabka commented:
> I don't think this patch fixes the following case:
> 
> ```c++
> template<typename T> 
> struct A 
> {
>     struct B 
>     {
>         using X = int;
> 
>         struct C
>         {
>             using X = void;
> 
>             struct D;
>         };
>     };
> };
> 
> template<typename T> 
> struct A<T>::B::C::D : B
> {
>     X x; // error: field has incomplete type 'X' (aka 'void')
> };
> ```
> 
> Regardless, this patch should probably include this as a test.

Many thanks for your answer! Yor're right, this test failed with this patch, but I've fixed it yet (see new changes). 

Moreover, I have added more tests to verify cases that should compile and those that shouldn't.

vbe-sc commented:
> I don't think this patch fixes the following case:
> 
> ```c++
> template<typename T> 
> struct A 
> {
>     struct B 
>     {
>         using X = int;
> 
>         struct C
>         {
>             using X = void;
> 
>             struct D;
>         };
>     };
> };
> 
> template<typename T> 
> struct A<T>::B::C::D : B
> {
>     X x; // error: field has incomplete type 'X' (aka 'void')
> };
> ```
> 
> Regardless, this patch should probably include this as a test.

Many thanks for your answer! You're right, this test failed with this patch, but I've already fixed it (see the new changes).
Moreover, I have added more tests that should and shouldn't compile.

cor3ntin commented on clang/lib/AST/CXXInheritance.cpp:
```diff
@@ -170,13 +170,21 @@ bool CXXBasePaths::lookupInBases(ASTContext &Context,
 
+    bool isCurrentInstantiation = isa<InjectedClassNameType>(BaseType);
+    if (!isCurrentInstantiation) {
+      if (auto *BaseRecord = cast_or_null<CXXRecordDecl>(
```
```suggestion
      if (auto *BaseRecord = cast_if_present<CXXRecordDecl>(
```

Can this actually be null?

vbe-sc replied:
Yes, you are right. Fixed here and in the other part of the code because there is the same case.

cor3ntin reviewed (commented):
Can you add a release note and update cxx_dr_status?
Thanks

vbe-sc commented:
> Can you add a release note

I added a release note as you asked

vbe-sc commented:
@cor3ntin could you, please, re-review this patch and resolve the comments if everything is ok?

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -268,6 +268,9 @@ Resolutions to C++ Defect Reports
   by default.
   (`CWG2521: User-defined literals and reserved identifiers <https://cplusplus.github.io/CWG/issues/2521.html>`_).
 
+- Clang now make correct name lookup when dependent base class is the current instantiation.
```
```suggestion
- Fix name lookup for a dependent base class that is the current instantiation.
```

vbe-sc replied:
Fixed, thanks

vbe-sc replied:
Could we merge this?

cor3ntin reviewed (approved):
LGTM (modulo small wording nit in the release note)

asi-sc commented:
@vbe-sc , reverted as you asked me offline https://github.com/llvm/llvm-project/pull/117727 . Feel free to create a new PR which includes the original commit and the fix for problem.

nikic commented:
FYI this seems to have significant cost for some types of C++ code, in particular clang build time regresses by 0.35% (https://llvm-compile-time-tracker.com/compare.php?from=4a7b56e6e7dd0f83c379ad06b6e81450bc691ba6&to=486644723038555a224fd09d462bb5099e64809e&stat=instructions:u).

erichkeane commented:
Hmm, that compile time regression is unfortunate.  @vbe-sc as a part of your next version of this patch, can you do some sort of analysis why this would result in further lookups/instantiation/etc?  I could comprehend that perhaps we're skipping the 'current instantiation' now for these bases so now there is more work to do in instantiations (of which there are more of those than the base, sort of out of necessity), but would like some sort of analysis/confirmation that is the case/what is happening in some of those 'worst' comparisons.

vbe-sc commented:
Such an example causes a failure in my patch

```c++
template <class T> 
class X {
public:
  X() = default;
  virtual ~X() = default;

  virtual int foo(int x, int y, T &entry) = 0;

  void bar() {
    struct Processor : public X<T> {
      Processor() : X() {}

      int foo(int, int, T &) override {
        return 42;
      }
    };
  }
};
```

The output 
```
llvm-project/clang/lib/AST/DeclCXX.cpp:2508: void clang::CXXMethodDecl::addOverriddenMethod(const CXXMethodDecl *): Assertion `!MD->getParent()->isDependentContext() && "Can't add an overridden method to a class template!"' failed.
```

The problem case is class definition nested to the method. 
I'm working on investigation and fixing this issue

vbe-sc commented:
```c++
void CXXMethodDecl::addOverriddenMethod(const CXXMethodDecl *MD) {
  assert(MD->isCanonicalDecl() && "Method is not canonical!");
  assert(!MD->getParent()->isDependentContext() &&
         "Can't add an overridden method to a class template!");
```

It seems to me that the second assertion is not really a requirement. @mizvekov , can you please comment on this statement?
